### ISSUE TITLE
GC overhead limit exceeded for spring-cloud-gcp

### ISSUE THREAD

**Issue by suztomo** [GoogleCloudPlatform/cloud-opensource-java#589]
To follow-up feedback from Mike.

------------------------

[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 02:56 min
[INFO] Finished at: 2019-04-29T13:00:28-04:00
[INFO] Final Memory: 134M/972M
[INFO] ------------------------------------------------------------------------
[ERROR] GC overhead limit exceeded -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/OutOfMemoryError

https://github.com/spring-cloud/spring-cloud-gcp/pull/1658 and run "mvn clean install -DskipTests".

**Comment by suztomo** [comment]
It fails with subproject "spring-cloud-gcp/spring-cloud-gcp-autoconfigure".

Heap showing many Artifact instances.
![image](https://user-images.githubusercontent.com/28604/56925620-e079d000-6a9d-11e9-9ac7-127e263e09eb.png)

Call tree showing listClassesInJar taking time
![image](https://user-images.githubusercontent.com/28604/56926066-d3111580-6a9e-11e9-8692-1e7c46f8654a.png)

**Comment by suztomo** [comment]
@meltsufin I checked the OOM. This enforcer rule takes more memory than Maven's default. So it needed to increase max heap memory in my environment. Thank you for reporting this.

# Solution

Increasing JVM heap size solved this.

```
 ~/git/spring-cloud/spring-cloud-gcp$ export MAVEN_OPTS='-Xmx2g'
 ~/git/spring-cloud/spring-cloud-gcp$ mvn clean install -DskipTests
...
[INFO] spring-cloud-gcp-kotlin-samples .................... SUCCESS [  0.196 s]
[INFO] spring-cloud-gcp-kotlin-app-sample ................. SUCCESS [ 14.063 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 08:38 min
[INFO] Finished at: 2019-04-29T17:20:08-04:00
[INFO] Final Memory: 222M/1979M
[INFO] ------------------------------------------------------------------------
```

![image](https://user-images.githubusercontent.com/28604/56928013-9b589c80-6aa3-11e9-99cb-852fc8fee7dc.png)

**Comment by elharo** [comment]
That's a surprising number of HashMaps. I wonder where they're coming from?

**Comment by suztomo** [comment]
Checking further.

**Comment by suztomo** [comment]
The previous image on heap was after OOM. Just before the OOM, the heap has many BCEL objects.

![image](https://user-images.githubusercontent.com/28604/56972361-d822a300-6b38-11e9-8c77-88500cfd868f.png)

The incoming references to one of the BCEL object shows it's retained by BCEL repository through hashmap.

![image](https://user-images.githubusercontent.com/28604/56972264-b0333f80-6b38-11e9-8eba-3ddd40b95b8e.png)

Digging further.

**Comment by suztomo** [comment]
ConstantUtf8 custom modification solved OutOfMemoryError.
![621k_constantutf8_with_getCachedInstance](https://user-images.githubusercontent.com/28604/57090951-10131d00-6cd6-11e9-98f4-c671e98c0a80.png)

The number of instances became 0.6 million from 2.6 million. Started email thread in dev@commons.apache.org

**Comment by suztomo** [comment]
Fixed as of 0.2.1 release.

```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Spring Cloud GCP 1.2.0.BUILD-SNAPSHOT:
[INFO] 
[INFO] Spring Cloud GCP ................................... SUCCESS [  9.825 s]
[INFO] Spring Cloud GCP Core Module ....................... SUCCESS [ 14.093 s]
[INFO] Spring Cloud GCP Stackdriver Logging Module ........ SUCCESS [ 11.191 s]
[INFO] Spring Cloud GCP Pub/Sub Module .................... SUCCESS [ 13.157 s]
[INFO] Spring Cloud GCP Storage Module .................... SUCCESS [  9.019 s]
[INFO] Spring Cloud GCP Datastore Module .................. SUCCESS [ 15.844 s]
[INFO] Spring Cloud GCP Cloud Spanner Module .............. SUCCESS [ 15.267 s]
[INFO] Spring Cloud GCP Vision Module ..................... SUCCESS [  8.831 s]
[INFO] Spring Cloud GCP IAP Authentication Module ......... SUCCESS [  9.715 s]
[INFO] Spring Cloud GCP Autoconfigure Module .............. SUCCESS [ 36.468 s]
[INFO] Spring Cloud GCP Starters .......................... SUCCESS [  0.113 s]
[INFO] Spring Cloud GCP Support Starter ................... SUCCESS [  4.989 s]
[INFO] Spring Cloud GCP Pub/Sub Starter ................... SUCCESS [  6.088 s]
[INFO] Spring Cloud GCP Cloud Firestore Starter ........... SUCCESS [  6.392 s]
[INFO] Spring Cloud GCP Documentation ..................... SUCCESS [  0.500 s]
[INFO] Spring Cloud GCP Dependencies ...................... SUCCESS [  0.015 s]
[INFO] Spring Cloud GCP Stream ............................ SUCCESS [ 17.741 s]
[INFO] Spring Cloud GCP Pub/Sub Bus Starter ............... SUCCESS [  9.309 s]
[INFO] Spring Cloud GCP Config Starter .................... SUCCESS [  5.025 s]
[INFO] Spring Cloud GCP Storage Starter ................... SUCCESS [  5.153 s]
[INFO] Spring Cloud GCP Spanner Starter ................... SUCCESS [  6.531 s]
[INFO] Spring Cloud GCP Datastore Starter ................. SUCCESS [  5.692 s]
[INFO] Spring Cloud GCP Cloud Trace Starter ............... SUCCESS [  7.892 s]
[INFO] Spring Cloud GCP Stackdriver Logging Starter ....... SUCCESS [  6.182 s]
[INFO] Spring Cloud GCP Cloud SQL MySQL Starter ........... SUCCESS [  6.538 s]
[INFO] Spring Cloud GCP Cloud SQL PostgreSQL Starter ...... SUCCESS [  5.585 s]
[INFO] Spring Cloud GCP Security IAP Starter .............. SUCCESS [  5.905 s]
[INFO] Spring Cloud GCP Cloud Vision Starter .............. SUCCESS [  6.806 s]
[INFO] Spring Cloud GCP Code Samples ...................... SUCCESS [  0.955 s]
[INFO] Spring Cloud GCP Runtime Configuration Code Sample . SUCCESS [ 12.330 s]
[INFO] Spring Cloud GCP Trace Code Sample ................. SUCCESS [ 21.024 s]
[INFO] Spring Cloud GCP Logging Code Sample ............... SUCCESS [  8.349 s]
[INFO] Spring Cloud GCP SQL MySQL Starter Code Sample ..... SUCCESS [ 13.076 s]
[INFO] Spring Cloud GCP Integration Pub/Sub Channel Adapters Code Sample SUCCESS [  0.095 s]
[INFO] Spring Cloud GCP Integration Pub/Sub Sender Code Sample SUCCESS [  9.491 s]
[INFO] Spring Cloud GCP Integration Pub/Sub Receiver Code Sample SUCCESS [  7.877 s]
[INFO] Spring Cloud GCP Integration Pub/Sub Polling Receiver Code Sample SUCCESS [  7.671 s]
[INFO] Spring Cloud GCP Integration Pub/Sub Integration Test SUCCESS [  9.736 s]
[INFO] Spring Cloud GCP Storage Sample .................... SUCCESS [ 13.095 s]
[INFO] Spring Cloud GCP Spanner Code Sample ............... SUCCESS [ 16.763 s]
[INFO] Spring Cloud GCP Datastore Code Sample ............. SUCCESS [ 14.945 s]
[INFO] Spring Cloud GCP Datastore Bookshelf Code Sample ... SUCCESS [ 10.246 s]
[INFO] Spring Cloud GCP Pub/Sub Code Sample ............... SUCCESS [  8.772 s]
[INFO] Spring Cloud GCP with Spring Data JPA Code Sample .. SUCCESS [ 11.199 s]
[INFO] Spring Cloud GCP Vision API Sample ................. SUCCESS [ 16.137 s]
[INFO] Spring Cloud GCP Integration Storage Sample ........ SUCCESS [  7.265 s]
[INFO] Spring Cloud GCP Stream Pub/Sub Code Sample ........ SUCCESS [  9.744 s]
[INFO] Spring Cloud GCP Pub/Sub Bus Configuration Management Code Sample SUCCESS [  0.091 s]
[INFO] Spring Cloud GCP Pub/Sub Bus Sample (Client) ....... SUCCESS [ 16.395 s]
[INFO] Spring Cloud GCP Pub/Sub Bus Sample (Local Configuration) SUCCESS [ 12.242 s]
[INFO] Spring Cloud GCP Pub/Sub Bus Sample (GitHub Configuration) SUCCESS [ 11.897 s]
[INFO] spring-cloud-gcp-pubsub-bus-config-sample-test ..... SUCCESS [  4.908 s]
[INFO] Spring Cloud GCP Stream Polling Pub/Sub Code Sample  SUCCESS [  9.644 s]
[INFO] Spring Cloud GCP Pub/Sub Reactive Code Sample ...... SUCCESS [ 10.587 s]
[INFO] Spring Cloud GCP Integration Pub/Sub JSON Payloads Code Sample SUCCESS [  9.205 s]
[INFO] Spring Cloud GCP Security IAP Sample ............... SUCCESS [ 14.494 s]
[INFO] spring-cloud-gcp-sql-postgres-sample ............... SUCCESS [ 13.149 s]
[INFO] Spring Cloud GCP Vision OCR Demo ................... SUCCESS [ 11.029 s]
[INFO] Spring Cloud GCP Firestore Code Sample ............. SUCCESS [ 12.788 s]
[INFO] spring-cloud-gcp-kotlin-samples .................... SUCCESS [  0.173 s]
[INFO] spring-cloud-gcp-kotlin-app-sample ................. SUCCESS [ 17.852 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  09:57 min
[INFO] Finished at: 2019-06-14T11:03:06-04:00
[INFO] ------------------------------------------------------------------------
```
